pkg/http/api.go

```diff
@@ -2053,6 +2074,115 @@ func (a *api) onGetOutboundHealthz(reqCtx *fasthttp.RequestCtx) {
 	}
 }
 
+func (a *api) getComponent(componentKind string, componentName string) (component interface{}) {
+	switch componentKind {
+	case "state":
+		if statestore := a.stateStores[componentName]; statestore != nil {
+			component = statestore
+		}
+	case "pubsub":
+		if pubsub := a.pubsubAdapter.GetPubSub(componentName); pubsub != nil {
+			component = pubsub
+		}
+	case "secretstores":
+		if secretstore := a.secretStores[componentName]; secretstore != nil {
+			component = secretstore
+		}
+	case "bindings":
+		if inputbinding := a.inputBindings[componentName]; inputbinding != nil {
+			component = inputbinding
+		} else if outputbinding := a.outputBindings[componentName]; outputbinding != nil {
+			component = outputbinding
+		}
+	}
+	return
+}
+
+func (a *api) onGetAllComponentsHealthz(reqCtx *fasthttp.RequestCtx) {
+	components := a.getComponentsFn()
+	hresp := ComponentsHealthResponse{
+		Results: make([]ComponentHealthItem, len(components)),
+	}
+	i := 0
+	for _, comp := range components {
+		a.allComponentsHealthResponsePopulator(strings.Split(comp.Spec.Type, ".")[0], hresp, i, comp.Name)
+		i++
+	}
+	b, _ := json.Marshal(hresp)
+	respond(reqCtx, withJSON(fasthttp.StatusOK, b))
+}
+
+func (a *api) allComponentsHealthResponsePopulator(componentType string, hresp ComponentsHealthResponse, ind int, name string) {
+	status, errStr, message := a.onGetComponentHealthzUtil(nil, componentType, name)
+	hresp.Results[ind].Component = name
+	hresp.Results[ind].Type = componentType
+	hresp.Results[ind].Status = status
+	if errStr != "" {
+		hresp.Results[ind].ErrorCode = errStr
+	}
+	if message != "" {
+		hresp.Results[ind].Message = message
+	}
+}
+
+func (a *api) onGetComponentHealthz(reqCtx *fasthttp.RequestCtx) {
+	componentName := reqCtx.UserValue(componentNameParam).(string)
+	components := a.getComponentsFn()
+	found := false
+	for _, comp := range components {
+		if componentName == comp.Name {
+			found = true
+			a.onGetComponentHealthzUtil(reqCtx, strings.Split(comp.Spec.Type, ".")[0], comp.Name)
+			break
+		}
+	}
+	if !found {
+		msg := NewComponentHealth(utils.StatusUndefined, messages.ErrComponentNotFound, "")
+		if reqCtx != nil {
+			respond(reqCtx, withHealthStatus(fasthttp.StatusBadRequest, msg))
+		}
+		log.Debug(msg)
+	}
+}
+
+func (a *api) onGetComponentHealthzUtil(reqCtx *fasthttp.RequestCtx, componentKind string,
+	componentName string,
+) (healthStatus string, errStr string, message string) {
+	component := a.getComponent(componentKind, componentName)
+
+	if component == nil {
+		msg := NewComponentHealth(utils.StatusUndefined, messages.ErrComponentNotFound, "")
+		if reqCtx != nil {
+			respond(reqCtx, withHealthStatus(fasthttp.StatusBadRequest, msg))
```

Comment:
Is there a way to refractor this function so that respond is only called from the request handler and not any of the dependent util functions that the handler calls... I belive this will lead to confusion....

Return only httpStatusCode, component Health  objects from this function and let the caller handle the response?

Reply:
Yeah, responding now from base function only